### DISCUSSION TITLE
Emotion & Chakra UI does not work with Rollup

### DISCUSSION THREAD

**Discussion by georgekrax** [chakra-ui/chakra-ui#4986] [Q&A]
Hello everyone,

I am trying to bundle my own React.js components and icons library, utilizing the amazing features of Chakra UI. I am using Rollup, and everything works fine. However, when I import my custom component to another project and use a theme or style prop, such as `bg="gray.500"`, this translates to CSS `background: gray.500` and not `background: var(--chakra-gray-500)`, as it was supposed to. 

> Note: When I use `tsc`, everything works fine, and the styles are "translated" to CSS variables.

So, I believe that Rollup.js is where my fault is located. When I add `external: ["@chakra-ui/icons"]` or `external: ["@chakra-ui/system"]`, to the `rollup.config.js` file, then everything also works great. 
Should I add and configure something with Babel.js in order to work with Emotion styles? Would that help?

The link to my repo is: https://github.com/hashtag-design-system/design_system/tree/dev/packages/icons

Any help or advice is appreciated!

@segunadebayo How you and your team bundle Chakra UI packages. Maybe you can help, as you are the people that work on this.

**Comment by justinphilpott** [comment]
Bump

**Reply by georgekrax** [reply]
What do you mean by "bumb"? Change version?

**Comment by justinphilpott** [comment]
Apologies, what I meant was, "please does anyone have any clues about this".

------- Original Message -------

**Reply by georgekrax** [reply]
Okay, that is better. Nice!

**Comment by sauldeleon** [comment]
any news on this? I am making a component library based on chakra and when doing the bundle with rollup and importing the components, the same issue is happening, theme properties are not being translated :(

@georgekrax did you solve this somehow?

**Reply by georgekrax** [reply]
Unfortunately, I do not remember what I did to fix the issue! I should have commented here on the solution the time I had it fixed. However, I do remember that I added Chakra - UI `peerDependencies` in the Rollup config file and `package.json`. Please see my solution below. 

> So, I believe that Rollup.js is where my fault is located. When I add external: ["@chakra-ui/icons"] or external: ["@chakra-ui/system"], to the rollup.config.js file, then everything also works great.


Another option in the `rollup.config.js` file is to config like, and then add in the `package.json` the `peerDependencies` as you wish/need.

`rollup.config.js`:
```js
import { peerDependencies } from "./package.json";

export default [
  {
    ...
    external: Object.keys(peerDependencies),
    ...
  },
];
```

`package.json`:
```json
"peerDependencies": {
    "@chakra-ui/react": "^1.8.7",
    "framer-motion": ">=6",
    "react": "^17.0.2"
  }
```
***The above configuration is for the main component package. But I use it also in the separate icons package**

`icons/package.json`:
```json
"peerDependencies": {
    "@chakra-ui/icons": "^1.1.7",
    "@chakra-ui/system": "^1.12.0",
    "react": "^17.0.2"
  }
```

Feel free to ask me, if you don't understand something! 😊

**Reply by HelixHEX** [reply]
Hey!

I'm using tsdx and I still can't get chakra-ui to work once it's packaged. It seems to through some sort of error

**Reply by georgekrax** [reply]
Unfortunately, I don't have any experience with `tsdx`. There could be many reasons. Do you use `@chakra-ui/icons`? Have you installed `@chakra-ui/react` on the project that you make use of your own custom package? Have you configured and used `@chakra-ui/react` **context** as it is expected?

I would advise you to have a look at [Chakra-UI's Discord Server](https://discord.com/channels/660863154703695893/695564068895064094), or send us a link of your repo.

**Reply by HelixHEX** [reply]
I originally used rollup and still had trouble. Also, tsdx uses rollup under the hood. 

I did have @chakra-ui/react installed,  don't use @chakra-ui/icons, and I'm in the discord.

**Reply by georgekrax** [reply]
I see. My `rollup.config.js` is this:
```js
import commonjs from "@rollup/plugin-commonjs";
import json from "@rollup/plugin-json";
import resolve from "@rollup/plugin-node-resolve";
import typescript from "@rollup/plugin-typescript";
// import typescript from 'rollup-plugin-typescript2';
// import bundleScss from "rollup-plugin-bundle-scss";
import dts from "rollup-plugin-dts";
import peerDepsExternal from "rollup-plugin-peer-deps-external";
import postcss from "rollup-plugin-postcss";
import { terser } from "rollup-plugin-terser";
import { main, module, types, peerDependencies } from "./package.json";
import { compilerOptions } from "./tsconfig.json";

const { outDir, declarationDir } = compilerOptions;

export default [
  {
    input: "src/index.tsx",
    output: [
      {
        file: main,
        format: "cjs",
        sourcemap: true,
      },
      {
        file: module,
        format: "esm",
        sourcemap: true,
      },
    ],
    external: Object.keys(peerDependencies),
    plugins: [
      peerDepsExternal(),
      json(), 
      resolve(),
      commonjs(),
      typescript({
        tsconfig: "./tsconfig.json",
        outDir: ".",
        declarationDir: declarationDir.replace(outDir + "/", ""),
        emitDeclarationOnly: true,
      }),
      // terser(),
    ],
  },
  {
    input: declarationDir + "/index.d.ts",
    output: [{ file: types, format: "es" }],
    plugins: [dts()],
  },
  // {
  //   input: "src/index.scss",
  //   output: [{ file: "build/hey.scss", format: "es" }],
  //   plugins: [bundleScss({ exclusive: false, output: "build/hey.scss" })],
  // },
  {
    input: "src/index.scss",
    output: { file: "build/index.css", format: "es" },
    plugins: [
      postcss({
        minimize: true,
        to: "build/index.css",
        extract: true,
        sourceMap: true,
      }),
    ],
  },
];
```

Make sure to pay attention to the `external: Object.keys(peerDependencies)` line, where Rollup also imports and **adds @chakra-ui/react** on the produced code. I don't know though, how this fits with *tsdx* 🙄

**Reply by HelixHEX** [reply]
Thanks so much for the insight. I'll try and update my roll up config file. 

tsdx allows for total customization when it comes to rollup so I can go in and change it how I best see fit

**Reply by georgekrax** [reply]
You're welcome. I didn't help a lot, because I have never used *tsdx* ever before.

**Reply by sachinkotwal** [reply]
Anyone found solution to this problem

**Reply by nick-inkeep** [reply]
+1

**Reply by dclawson** [reply]
+1

**Comment by makeyzn** [comment]
For the future, this answer on Stack Overflow helped me solve this problem
https://stackoverflow.com/questions/75627098/styles-not-showing-up-with-bundles-up-chakra-ui-components-using-rollup-with-ty?newreg=e13a466ec2e54babb2b79a36d6217d02
